### PR TITLE
Gate the usage of nested multisigs on a block version change

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

We'd like to ensure that nested multisigs (in mint governors and
MintConfigTxs) only start showing up in the blockchain once clients have
had sufficient time to upgrade.

Similarly to other blockchain changes, we achieve this by gating the new
feature on a block version upgrade.

### CODE REVIEW

**Comment by eranrund** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2732** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2732" target="_blank">Graphite</a> 
    * **PR #2734** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2734" target="_blank">Graphite</a> 
      * **PR #2736** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2736" target="_blank">Graphite</a> 
        * **PR #2767** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2767" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/2767?utm_source=stack-comment).

**Review by garbageslam — APPROVED** [review]

File: consensus/enclave/impl/src/lib.rs [hunk 1]
```diff
@@ -1127,6 +1143,49 @@ mod tests {
         );
     }
 
+    #[test_with_logger]
+    fn test_enclave_init_refuses_nested_multisig_governors_when_not_supported(logger: Logger) {
```
Comment (nick-mobilecoin) [inline consensus/enclave/impl/src/lib.rs:1147]: :thought_balloon: This is sufficient for testing that `enclave_init()` obeys the governor validation logic, but it feels like the governor map validation logic itself could be more robustly tested.  This goes back to the suggestion of moving the validation to it's own function where one can focus on `BlockChainConfig`.
Some boundary cases:
- Empty map (this is exercised but it's not intentional and hard to locate in the tests)
- map with more than one item
  - all flat
  - First with multi
  - a middle entry has multi
  - the last element multi

Especially if the validation logic is moved to it's own function these cases should be able to be concise.
Though these test cases seem trivial and can be easily said to work by looking at the code, the intent is to provide guard rails for future maintainers.  For instance the `mint_trust_root_public_key` could get refactored to use the method from the Enclave structure (I would suggest updating to a freestanding function).  Without robust tests around this trivial logic (there is currently a test) someone can't refactor with confidence that they aren't inadvertently breaking something.

**Review by nick-mobilecoin — APPROVED** [review]

File: consensus/enclave/impl/src/lib.rs [hunk 2]
```diff
@@ -475,6 +475,22 @@ impl ConsensusEnclave for SgxConsensusEnclave {
             minting_trust_root_public_key
                 .verify_governors_map(&blockchain_config.governors_map, &signature)
                 .map_err(|_| Error::InvalidGovernorsSignature)?;
+
```
[the remark this answers is not in the document]
Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs:478]: Good call, I'll break it out.

File: transaction/core/src/mint/validation/config.rs [hunk 3]
```diff
@@ -130,20 +146,44 @@ mod tests {
...
+        }
     }
 
     #[test]
     fn validate_configs_accepts_no_configs() {
-        assert!(validate_configs(123.into(), &[]).is_ok());
+        for block_version in BlockVersion::iterator() {
+            if !block_version.mint_transactions_are_supported() {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/mint/validation/config.rs:183]: I don't agree with that, since if we end up moving the block version at which we release this feature, we're gonna need to remember to go and bump the number in all these tests...
I don't think we care too much about the specific block version number, we just want to test all versions that happen to support mint transactions. For example, it is possible that at some point in the future they will get deprecated and we will stop supporting them and for some period the code will continue to exist (especially during the transition period). If we use straight up block version numbers we'll then need to also limit to the upper bound, and again hardcode a number in a bunch of tests. I think it's fairly clear to the reader/future maintainer what is happening here.

File: transaction/core/src/mint/validation/config.rs [hunk 4]
```diff
@@ -194,14 +248,49 @@ mod tests {
...
+                || block_version.nested_multisigs_are_supported()
+            {
+                continue;
+            }
+
+            assert_eq!(
+                validate_configs(123.into(), &[mint_config.clone()], block_version),
+                Err(Error::InvalidSignerSet)
+            );
+        }
```
[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/mint/validation/config.rs:293]: Unfortunately it's not yet guaranteed that this is the only version, since if this doesn't make it into the next release there might be two versions that did not support nesting. Hence why I prefer to rely on the specific properties of each version.

File: transaction/core/src/mint/validation/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/mint/validation/config.rs:183]: (with that said, I don't feel strongly about this so if `.skip(2)` gets this to move quicker down the review pipeline I am happy to switch to that)

File: consensus/enclave/impl/src/lib.rs (hunk 1 above)
Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs:1121]: Hm, the map here is a hashmap of `TokenId` -> `SignerSet` so not sure what you mean by first with multi, middle entry with multi, etc. The `SignerSet` has two separate lists - one for individual pubkeys and one for `SignerSet`s.

File: transaction/core/src/mint/validation/config.rs [hunk 5]
```diff
@@ -130,20 +146,44 @@ mod tests {
...
+            }
+
+            let mut configs = vec![
+                mint_config1.clone(),
+                mint_config2.clone(),
+                mint_config3.clone(),
+            ];
+            if block_version.nested_multisigs_are_supported() {
+                configs.push(mint_config4.clone());
+            }
```
[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/mint/validation/config.rs:171]: I'll break it into a separate test

File: transaction/core/src/mint/validation/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply transaction/core/src/mint/validation/config.rs:199]: I can buy the argument of avoiding hardcoding block versions to minimize test maintenance. (This and other comments)

[the remark this answers is not in the document]
Comment (eranrund) [reply transaction/core/src/mint/validation/config.rs:198]: If that feels good to you then I will mark the comments related to this as resolved.

File: consensus/enclave/impl/src/lib.rs (hunk 1 above)
Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs]: How do you feel about the [revised tests](https://github.com/mobilecoinfoundation/mobilecoin/blob/6d02f913c2532715f206e8fb91441d51654b0701/consensus/enclave/api/src/config.rs#L318-L493)?

Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs]: Ping @nick-mobilecoin

File: consensus/enclave/api/src/config.rs [hunk 6]
```diff
@@ -354,4 +367,128 @@ mod test {
             Err(Error::InvalidGovernorsSignature)
         );
     }
+
+    #[test]
+    fn validate_accepts_nested_multisig_govenors_when_supported() {
```
Comment (nick-mobilecoin) [inline consensus/enclave/api/src/config.rs:372]: :wrench: 
```suggestion
    fn validate_accepts_nested_multisig_governors_when_supported() {
```

File: consensus/enclave/api/src/config.rs [hunk 7]
```diff
@@ -354,4 +367,128 @@ mod test {
...
+                governors_signature,
+                block_version,
+            };
+
+            assert_eq!(config.validate(&governors_public_key), Ok(()));
+        }
+    }
+
+    #[test]
+    fn validate_accepts_nested_multisig_with_individual_govenors_when_supported() {
```
Comment (nick-mobilecoin) [inline consensus/enclave/api/src/config.rs:400]: :wrench: 
```suggestion
    fn validate_accepts_nested_multisig_with_individual_governors_when_supported() {
```

File: consensus/enclave/api/src/config.rs [hunk 8]
```diff
@@ -354,4 +367,128 @@ mod test {
...
+                governors_signature,
+                block_version,
+            };
+
+            assert_eq!(config.validate(&governors_public_key), Ok(()));
+        }
+    }
+
+    #[test]
+    fn validate_rejects_nested_multisig_govenors_when_not_supported() {
```
Comment (nick-mobilecoin) [inline consensus/enclave/api/src/config.rs:428]: :wrench: 
```suggestion
    fn validate_rejects_nested_multisig_governors_when_not_supported() {
```

File: consensus/enclave/api/src/config.rs [hunk 9]
```diff
@@ -354,4 +367,128 @@ mod test {
...
+                Err(Error::NestedMultiSigGovernorsNotSupported(
+                    TokenId::from(1),
+                    block_version
+                ))
+            );
+        }
+    }
+
+    #[test]
+    fn validate_rejects_nested_multisig_with_individual_govenors_when_not_supported() {
```
Comment (nick-mobilecoin) [inline consensus/enclave/api/src/config.rs:462]: :wrench: 
```suggestion
    fn validate_rejects_nested_multisig_with_individual_governors_when_not_supported() {
```

**Review by nick-mobilecoin — APPROVED** [review]

File: consensus/enclave/api/src/config.rs [hunk 10]
```diff
@@ -354,4 +367,128 @@ mod test {
...
+                vec![SignerSet::new(vec![Ed25519Public::default()], 1)],
+                1,
+            ),
+        )])
+        .unwrap();
+
+        let (governors_signature, governors_public_key) = sign_governors_map(&nested_governors_map);
+
+        for block_version in
+            BlockVersion::iterator().filter(|bv| bv.nested_multisigs_are_supported())
```
[the remark this answers is not in the document]
Comment (eranrund) [reply consensus/enclave/api/src/config.rs:414]: It does look a bit nicer even if the functionality is the same :)

**Review by nick-mobilecoin — APPROVED** [review]

File: consensus/enclave/impl/src/lib.rs (hunk 1 above)
Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs]: No problem! Thanks for the reviews.
I am not seeing any new nits though...
